pyomo/contrib/solver/solvers/knitro/utils.py

```diff
@@ -0,0 +1,272 @@
+
+    """
+    for var in variables:
+        i = var_map[id(var)]
+        var.set_value(values[i])
+
+
+class NonlinearExpressionData(Function):
+    """Holds the data required to evaluate a non-linear expression.
+
+    This class stores a Pyomo expression along with its variables and can compute
+    gradient and Hessian information for use with optimization solvers.
+
+    Attributes:
+        variables (list[VarData]): list of variables referenced in the expression.
+        func_expr (Expression): The Pyomo expression representing the non-linear function.
+        grad_map (Mapping[VarData, Expression]): Gradient expressions mapped by variable.
+        hess_map (Mapping[tuple[VarData, VarData], Expression]): Hessian expressions
+            mapped by variable pairs.
+        diff_order (int): Level of differentiation to compute:
+            - 0: function evaluation only
+            - 1: function + gradient
+            - 2: function + gradient + hessian
+
+    """
+
+    variables: list[VarData]
+    func_expr: Expression
+    grad_map: Mapping[VarData, Expression]
+    hess_map: Mapping[tuple[VarData, VarData], Expression]
+    diff_order: int
+
+    def __init__(
+        self,
+        expr: Expression,
+        variables: Iterable[VarData],
+        var_map: Mapping[int, int],
+        diff_order: int = 0,
+    ) -> None:
+        """Initialize NonlinearExpressionData.
+
+        Args:
+            expr (Expression): The Pyomo expression to evaluate.
+            variables (Iterable[VarData]): Variables referenced in the expression.
+            diff_order (int): Level of differentiation to compute:
+                - 0: function evaluation only
+                - 1: function + gradient
+                - 2: function + gradient + hessian
+
+        """
+        self.func_expr = expr
+        self.variables = list(variables)
+        self.diff_order = diff_order
+        self._var_map = var_map
+        if diff_order >= 1:
+            self.compute_gradient()
+        if diff_order >= 2:
+            self.compute_hessian()
+
+    @property
+    def grad_vars(self) -> list[VarData]:
+        """Get the list of variables for which gradients are available.
+
+        Returns:
+            list[VarData]: Variables with gradient information.
+
+        """
+        return list(self.grad_map.keys())
+
+    @property
+    def hess_vars(self) -> list[tuple[VarData, VarData]]:
+        """Get the list of variable pairs for which Hessian entries are available.
+
+        Returns:
+            list[tuple[VarData, VarData]]: Variable pairs with Hessian information.
+
+        """
+        return list(self.hess_map.keys())
+
+    def compute_gradient(self) -> None:
+        """Compute gradient expressions for the nonlinear expression.
+
+        This method computes the gradient of the expression with respect to all
+        variables and stores the results in the grad attribute.
+        """
+        derivative = reverse_sd(self.func_expr)
```

Comment:
TODO: we should evaluate the computational performance of native python symbolic derivatives.  Some alternatives:
- compile the expressions into executable python code (to avoud needing to use the Pyomo expression evaluator
- use AD instead of SD (particularly the APPSI c-extension?)